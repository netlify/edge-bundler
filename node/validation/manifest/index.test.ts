/* eslint-disable max-nested-callbacks */
import { test, expect, describe } from 'vitest'

import { validateManifest, ManifestValidationError } from './index.js'

// Factory so we have a new object per test
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getBaseManifest = (): Record<string, any> => ({
  bundles: [
    {
      asset: 'f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js',
      format: 'js',
    },
  ],
  routes: [
    {
      name: 'name',
      function: 'hello',
      pattern: '^/hello/?$',
    },
  ],
  post_cache_routes: [
    {
      name: 'name',
      function: 'hello',
      pattern: '^/hello/?$',
    },
  ],
  layers: [
    {
      flag: 'flag',
      name: 'name',
      local: 'local',
    },
  ],
  bundler_version: '1.6.0',
})

test('should not throw on valid manifest', () => {
  expect(() => validateManifest(getBaseManifest())).not.toThrowError()
})

test('should throw ManifestValidationError with correct message', () => {
  expect(() => validateManifest('manifest')).toThrowError(ManifestValidationError)
  expect(() => validateManifest('manifest')).toThrowError(/^Validation of Edge Functions manifest failed/)
})

test('should throw on additional property on root level', () => {
  const manifest = getBaseManifest()
  manifest.foo = 'bar'

  expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
})

test('should show multiple errors', () => {
  const manifest = getBaseManifest()
  manifest.foo = 'bar'
  manifest.baz = 'bar'

  expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
})

describe('bundle', () => {
  test('should throw on additional property in bundle', () => {
    const manifest = getBaseManifest()
    manifest.bundles[0].foo = 'bar'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing asset', () => {
    const manifest = getBaseManifest()
    delete manifest.bundles[0].asset

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing format', () => {
    const manifest = getBaseManifest()
    delete manifest.bundles[0].format

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on invalid format', () => {
    const manifest = getBaseManifest()
    manifest.bundles[0].format = 'foo'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})
describe('route', () => {
  test('should throw on additional property', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].foo = 'bar'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on invalid pattern', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].pattern = '/^/hello/?$/'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing function', () => {
    const manifest = getBaseManifest()
    delete manifest.routes[0].function

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing pattern', () => {
    const manifest = getBaseManifest()
    delete manifest.routes[0].pattern

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})

// No tests for post_cache_routes as schema shared with routes

describe('layers', () => {
  test('should throw on additional property', () => {
    const manifest = getBaseManifest()
    manifest.layers[0].foo = 'bar'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing name', () => {
    const manifest = getBaseManifest()
    delete manifest.layers[0].name

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing flag', () => {
    const manifest = getBaseManifest()
    delete manifest.layers[0].flag

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})
/* eslint-enable max-nested-callbacks */
