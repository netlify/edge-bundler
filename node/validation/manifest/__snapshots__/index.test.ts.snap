// Vitest Snapshot v1

exports[`bundle > should throw on additional property in bundle 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  4 |       \\"asset\\": \\"f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js\\",
  5 |       \\"format\\": \\"js\\",
> 6 |       \\"foo\\": \\"bar\\"
    |       ^^^^^ 😲  foo is not expected to be here!
  7 |     }
  8 |   ],
  9 |   \\"routes\\": ["
`;

exports[`bundle > should throw on invalid format 1`] = `
"Validation of Edge Functions manifest failed
ENUM must be equal to one of the allowed values
(eszip2, js)

  3 |     {
  4 |       \\"asset\\": \\"f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js\\",
> 5 |       \\"format\\": \\"foo\\"
    |                 ^^^^^ 👈🏽  Unexpected value, should be equal to one of the allowed values
  6 |     }
  7 |   ],
  8 |   \\"routes\\": ["
`;

exports[`bundle > should throw on missing asset 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'asset'

  1 | {
  2 |   \\"bundles\\": [
> 3 |     {
    |     ^ ☹️  asset is missing here!
  4 |       \\"format\\": \\"js\\"
  5 |     }
  6 |   ],"
`;

exports[`bundle > should throw on missing format 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'format'

  1 | {
  2 |   \\"bundles\\": [
> 3 |     {
    |     ^ ☹️  format is missing here!
  4 |       \\"asset\\": \\"f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js\\"
  5 |     }
  6 |   ],"
`;

exports[`import map URL > should throw on wrong type 1`] = `
"Validation of Edge Functions manifest failed
TYPE must be string

  28 |   ],
  29 |   \\"bundler_version\\": \\"1.6.0\\",
> 30 |   \\"importMapURL\\": [
     |                   ^
> 31 |     \\"file:///root/.netlify/edge-functions-dist/import_map.json\\"
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 32 |   ]
     | ^^^^ 👈🏽  type must be string
  33 | }"
`;

exports[`layers > should throw on additional property 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  25 |       \\"name\\": \\"name\\",
  26 |       \\"local\\": \\"local\\",
> 27 |       \\"foo\\": \\"bar\\"
     |       ^^^^^ 😲  foo is not expected to be here!
  28 |     }
  29 |   ],
  30 |   \\"bundler_version\\": \\"1.6.0\\""
`;

exports[`layers > should throw on missing flag 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'flag'

  21 |   ],
  22 |   \\"layers\\": [
> 23 |     {
     |     ^ ☹️  flag is missing here!
  24 |       \\"name\\": \\"name\\",
  25 |       \\"local\\": \\"local\\"
  26 |     }"
`;

exports[`layers > should throw on missing name 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'name'

  21 |   ],
  22 |   \\"layers\\": [
> 23 |     {
     |     ^ ☹️  name is missing here!
  24 |       \\"flag\\": \\"flag\\",
  25 |       \\"local\\": \\"local\\"
  26 |     }"
`;

exports[`route > should throw on additional property 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  11 |       \\"function\\": \\"hello\\",
  12 |       \\"pattern\\": \\"^/hello/?$\\",
> 13 |       \\"foo\\": \\"bar\\"
     |       ^^^^^ 😲  foo is not expected to be here!
  14 |     }
  15 |   ],
  16 |   \\"post_cache_routes\\": ["
`;

exports[`route > should throw on invalid pattern 1`] = `
"Validation of Edge Functions manifest failed
FORMAT pattern needs to be a regex that starts with ^ and ends with $ without any additional slashes before and afterwards

  10 |       \\"name\\": \\"name\\",
  11 |       \\"function\\": \\"hello\\",
> 12 |       \\"pattern\\": \\"/^/hello/?$/\\"
     |                  ^^^^^^^^^^^^^^ 👈🏽  format pattern needs to be a regex that starts with ^ and ends with $ without any additional slashes before and afterwards
  13 |     }
  14 |   ],
  15 |   \\"post_cache_routes\\": ["
`;

exports[`route > should throw on missing function 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'function'

   7 |   ],
   8 |   \\"routes\\": [
>  9 |     {
     |     ^ ☹️  function is missing here!
  10 |       \\"name\\": \\"name\\",
  11 |       \\"pattern\\": \\"^/hello/?$\\"
  12 |     }"
`;

exports[`route > should throw on missing pattern 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'pattern'

   7 |   ],
   8 |   \\"routes\\": [
>  9 |     {
     |     ^ ☹️  pattern is missing here!
  10 |       \\"name\\": \\"name\\",
  11 |       \\"function\\": \\"hello\\"
  12 |     }"
`;

exports[`should show multiple errors 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  28 |   ],
  29 |   \\"bundler_version\\": \\"1.6.0\\",
> 30 |   \\"foo\\": \\"bar\\",
     |   ^^^^^ 😲  foo is not expected to be here!
  31 |   \\"baz\\": \\"bar\\"
  32 | }

ADDTIONAL PROPERTY must NOT have additional properties

  29 |   \\"bundler_version\\": \\"1.6.0\\",
  30 |   \\"foo\\": \\"bar\\",
> 31 |   \\"baz\\": \\"bar\\"
     |   ^^^^^ 😲  baz is not expected to be here!
  32 | }"
`;

exports[`should throw on additional property on root level 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  28 |   ],
  29 |   \\"bundler_version\\": \\"1.6.0\\",
> 30 |   \\"foo\\": \\"bar\\"
     |   ^^^^^ 😲  foo is not expected to be here!
  31 | }"
`;
